core/src/software/aws/toolkits/core/utils/Waiter.kt

```diff
@@ -3,14 +3,87 @@
 
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
 import software.amazon.awssdk.awscore.exception.AwsServiceException
 import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.CompletionException
+import kotlin.math.min
+import kotlin.reflect.KClass
+
+object Waiters {
+    private val LOG = getLogger<Waiters>()
+
+    /**
+     * Generic waiter method to wait for a certain AWS resource to be in a steady status until it reaches in a failed status
+     * or time out for pulling.
+     * @param T The AWS resource type for querying the status from it
+     */
+    fun <T> waitUntil(
+        // The success predicate based on the returned resource
+        succeedOn: (T) -> Boolean = { true },
+        // The fail predicate based on the returned resource. Return an error message if it fails, null otherwise
+        failOn: (T) -> Boolean = { false },
+        exceptionsToStopOn: Set<KClass<out Exception>> = emptySet(),
+        exceptionsToIgnore: Set<KClass<out Exception>> = emptySet(),
+        maxDuration: Duration = Duration.ofMinutes(1),
+        // The status pulling method to get the latest resource
+        call: () -> T
+    ): T? {
+        val start = Instant.now()
+        var attempt = 0
+
+        return runBlocking {
+            while (Duration.between(start, Instant.now()) < maxDuration) {
+                attempt++
+
+                try {
+                    val result = call()
+                    if (succeedOn.invoke(result)) {
+                        LOG.info { "Got expected response: $result" }
+                        return@runBlocking result
+                    }
+
+                    if (failOn.invoke(result)) {
+                        throw WaiterUnrecoverableException("Received a response that matched the failOn predicate: $result")
+                    }
+
+                    LOG.info { "Attempt $attempt failed predicate." }
+                } catch (e: WaiterUnrecoverableException) {
+                    throw e
+                } catch (e: Exception) {
+                    val cause = if (e is CompletionException) e.cause ?: e else e
+
+                    if (cause::class in exceptionsToStopOn) {
+                        LOG.info { "Got expected exception: ${cause::class}" }
+                        return@runBlocking null
+                    }
+
+                    if (cause::class in exceptionsToIgnore) {
+                        LOG.info { "Attempt $attempt failed with an expected exception (${cause::class})" }
+                    } else {
+                        throw e
+                    }
+                }
+
+                delay(calculateBackOff(attempt))
+            }
+
+            throw WaiterTimeoutException("Condition was not met after $attempt attempts (${Duration.between(start, Instant.now()).seconds} seconds)")
+        }
+    }
+
+    private fun calculateBackOff(attempt: Int) = 250L shl min(attempt - 1, 4) // Max = 250 * 2^4 = 4_000.
```

Comment:
That's the max, but what strategy is captured here? It looks like exponential backoff? also backoffms?

Reply:
I stole it from the Java SDK test util waiter...its exponential backoff

Reply:
Right but can we add a comment about it + that it's in MS